increase efficiency of line buffered I/O

grondo opened the pull request:
Fixes #2287 

This PR includes the addition of `flux_buffer_has_lines()` and its use in place of `flux_buffer_lines()` as discussed in #2287.

Additionally, the "multiple callback" support for `flux_buffer_t` read callbacks was removed -- while it makes sense for standalone use, in the larger context of use by libsubprocess in a reactor, repeatedly calling the buffer callback until no more data was consumed could result in other parts of the system getting starved out by a heavily used buffer.

Since the multiple callback support was removed, I just removed the tests for that aspect of the code. Hope that was an ok approach.

Here's some timing results, before:
```console
Executing: git log --oneline -n1; make -j 12 >/dev/null 2>&1 && time src/cmd/flux start flux srun t/shell/lptest 79 10000 | wc -l
a972f14 libflux/test: remove multi-callback tests for flux_buffer_t
10000

real	0m56.864s
user	0m56.344s
sys	0m3.962s
```

After:
```console
Executing: git log --oneline -n1; make -j 12 >/dev/null 2>&1 && time src/cmd/flux start flux srun t/shell/lptest 79 10000 | wc -l
177cd4a libflux/test: test flux_buffer_has_line
10000

real	0m4.796s
user	0m2.259s
sys	0m3.057s
```

grondo commented:
@chu11 suggested we should also add tests to ensure `flux_buffer_read/read_line` can be called multiple times from the read callbacks. That's a good idea, so I'll add that to the buffer unit tests.

grondo commented:
Ok, just pushed a test to ensure that `flux_buffer_t` read callbacks can call `read` or `read_line` multiple times before returning.

garlick commented:
This all looks good to me but methinks @chu11 should press the button after he's had a look.

grondo commented:
> just one comment on the signature of flux_buffer_has_line(). I used an integer return with -1 on error, 1 on true, 0 on false for flux_buffer_is_readonly(). So for consistency, would we want to copy that function's semantic? Or copy this bool return to flux_buffer_is_readonly()?

Good comment. Sorry I didn't look at existing usage. I'll just quickly change `flux_buffer_is_readonly()` if it is ok with you. It just seems like trying to handle errors with these functions is overkill. They are used in conjunction with other calls where the error can be caught.

chu11 commented:
>  I'll just quickly change flux_buffer_is_readonly() if it is ok with you.

Sounds good!

grondo commented:
Ok, I did a quick change to `flux_buffer_is_readonly()`. It still sets errno on an invalid buffer, so an error condition can be checked by setting `errno = 0` before the call and checking the value of errno after the call (though not sure that is needed, it does allow unit testing)

I changed `flux_buffer_has_line()` to use this same scheme (and squashed since the change was very small)

grondo commented:
Well, unfortunately travis seems to be misbehaving at the moment, will check again in the morning.

chu11 commented:
With the update to `flux_buffer_is_readonly()`, I think an update to `subprocess_read()` in libsubprocess needs to be updated so the `readonly` variable is a bool instead of an int.  Then there is a caller in `flux_subprocess_getline()`.

I think the other calls to is_readonly are fine as is.

grondo commented:
Thanks @chu11! I made the suggested changes. I was being overcareful about changing the libsubprocess api and didn't even look long enough to realize those users were internal.

Just to verify, it is fine to leave return type of `flux_subprocess_read_stream_closed()` as  `int`?

chu11 commented:
> Just to verify, it is fine to leave return type of flux_subprocess_read_stream_closed() as int?

I had debated if we should change that function signature too, but since it's not inconsistent to any other functions in libsubprocess, I figure we could leave it.  If you have time and want to update, that'd be fine too.  Just LMK if I should hit the button or if you wanna sneak in that one signature change too.

grondo commented:
I was hesitant to change the public libsubprocess API, I think we can leave
it for now.

On Thu, Aug 8, 2019, 7:50 AM Al Chu <notifications@github.com> wrote:

> Just to verify, it is fine to leave return type of
> flux_subprocess_read_stream_closed() as int?
>
> I had debated if we should change that function signature too, but since
> it's not inconsistent to any other functions in libsubprocess, I figure we
> could leave it. If you have time and want to update, that'd be fine too.
> Just LMK if I should hit the button or if you wanna sneak in that one
> signature change too.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/2294?email_source=notifications&email_token=AAFVEUSKGQ7XXXYXWQTJBPDQDQXCPA5CNFSM4IKERSZ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD333YCI#issuecomment-519552009>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFVEURXTYVNYNJI75E7CZDQDQXCPANCNFSM4IKERSZQ>
> .
>

chu11 commented:
> I was hesitant to change the public libsubprocess API, I think we can leave it for now.

Sounds good, merging.  Thanks!

grondo commented:
Thanks!
